src/cmd/ksh93/sh/io.c

```diff
@@ -1753,18 +1753,19 @@ void sh_iosave(Shell_t *shp, int origfd, int oldtop, char *name) {
             __builtin_unreachable();
         }
         if (savefd < 0 && (sp = shp->sftable[origfd]) && (sfset(sp, 0, 0) & SF_STRING)) {
-            const char *devnull = "/dev/null";
             savestr = 1;
-            fd = open(devnull, O_RDONLY | O_CLOEXEC);
+            int fd = open("/dev/null", O_RDONLY | O_CLOEXEC);
 
-            if (fd < 0) {
-                errormsg(SH_DICT, ERROR_system(1), e_open, devnull);
+            if (fd == -1) {
+                errormsg(SH_DICT, ERROR_system(1), e_open, "/dev/null");
                 __builtin_unreachable();
             }
 
             if (fd < 10) {
                 savefd = sh_fcntl(fd, F_DUPFD_CLOEXEC, 10);
                 close(fd);
+            } else {
+                savefd = fd;
```

Comment:
I am not sure how to verify this fix is correct.

Reply:
I don't know how to test this since it clearly requires two unusual conditions. But it is fairly obvious that the intent is to ensure `savefd` is greater than 9 when the subsequent code is executed. So if the preceding `open()` assigns a fd >= 10 there isn't any need for the `sh_fcntl` to "move" the fd.